### PR TITLE
Index AnVIL `diagnoses` entity (#4618)

### LINKED ISSUE #4618: Index AnVIL `diagnoses` entity

### PR DESCRIPTION
**Pull request by nadove-ucsc**
Connected issues: #4618

### Author

- [x] PR is a draft
- [x] Target branch is `develop`
- [x] Name of PR branch matches `issues/<GitHub handle of author>/<issue#>-<slug>`
- [x] PR title references all connected issues
- [x] PR title matches<sup>1</sup> that of a connected issue <sub>or comment in PR explains why they're different</sub>
- [x] For each connected issue, there is at least one commit whose title references that issue
- [x] PR is connected to all connected issues via ZenHub
- [x] PR description links to connected issues
- [x] Added `partial` label to PR <sub>or this PR completely resolves all connected issues</sub>

<sup>1</sup> when the issue title describes a problem, the corresponding PR
title is `Fix: ` followed by the issue title

### Author (reindex, API changes)

- [x] Added `r` tag to commit title <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR <sub>or this PR does not require reindexing</sub>
- [x] Added `a` (compatible changes) or `A` (incompatible ones) tag to commit title <sub>or this PR does not modify the Azul service API</sub>
- [x] Added `API` label to connected issues <sub>or this PR does not modify the Azul service API</sub>

### Author (chains)

- [x] This PR is blocked by previous PR in the chain <sub>or this PR is not chained to another PR</sub>
- [x] Added `base` label to the blocking PR <sub>or this PR is not chained to another PR</sub>
- [x] Added `chained` label to this PR <sub>or this PR is not chained to another PR</sub>

### Author (upgrading)

- [x] Documented upgrading of deployments in UPGRADING.rst <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR <sub>or this PR does not require upgrading</sub>

### Author (operator tasks)

- [x] Added checklist items for additional operator tasks <sub>or this PR does not require additional tasks</sub>

### Author (hotfixes)

- [x] Added `F` tag to main commit title <sub>or this PR does not include permanent fix for a temporary hotfix</sub>
- [x] Reverted the temporary hotfixes for any connected issues <sub>or the `prod` branch has no temporary hotfixes for any connected issues</sub>

### Author (before every review)

- [x] Rebased PR branch on `develop`, squashed old fixups
- [x] Ran `make requirements_update` <sub>or this PR does not touch requirements*.txt, common.mk, Makefile and Dockerfile</sub>
- [x] Added `R` tag to commit title <sub>or this PR does not touch requirements*.txt</sub>
- [x] Added `reqs` label to PR <sub>or this PR does not touch requirements*.txt</sub>
- [x] `make integration_test` passes in personal deployment <sub>or this PR does not touch functionality that could break the IT</sub>

### Peer reviewer (after requesting changes)

Uncheck the *Author (before every review)* checklists.

### Peer reviewer (after approval)

- [x] PR is not a draft
- [x] Ticket is in *Review requested* column
- [x] Requested review from primary reviewer
- [x] Assigned PR to primary reviewer

### Primary reviewer (after requesting changes)

Uncheck the *before every review* checklists. Update the `N reviews` label.

### Primary reviewer (after approval)

- [x] Actually approved the PR
- [x] Labeled connected issues as `demo` or `no demo`
- [x] Commented on connected issues about demo expectations <sub>or all connected issues are labeled `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] PR title is appropriate as title of merge commit
- [x] `N reviews` label is accurate
- [x] Moved ticket to *Approved* column
- [x] Assigned PR to current operator

### Operator (before pushing merge the commit)

Note for operator: only `anvilbox` and `anvildev` need to be reindexed.

- [x] Checked `reindex` label and `r` commit title tag
- [x] Checked that demo expectations are clear <sub>or all connected issues are labeled `no demo`</sub>
- [x] Squashed PR branch and rebased onto `develop`
- [x] Sanity-checked history
- [x] Pushed PR branch to GitHub
- [x] Pushed PR branch to GitLab `dev` and added `sandbox` label <sub>or PR is labeled `no sandbox`</sub>
- [x] Pushed PR branch to GitLab `anvildev` <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in `sandbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in `anvilbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Reviewed build logs for anomalies in `sandbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Reviewed build logs for anomalies in `anvilbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Deleted unreferenced indices in `sandbox` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Deleted unreferenced indices in `anvilbox` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Started reindex in `sandbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Started reindex in `anvilbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `sandbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `anvilbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Added PR reference to merge commit title
- [x] Collected commit title tags in merge commit title
- [x] Moved connected issues to Merged column in ZenHub
- [x] Pushed merge commit to GitHub

### Operator (after pushing the merge commit)

- [x] Shortened the PR chain <sub>or this PR is not labeled `base`</sub>
- [x] Pushed merge commit to GitLab `dev` <sub>or PR is labeled `no sandbox`</sub>
- [x] Pushed merge commit to GitLab `anvildev` <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes on GitLab `dev`<sup>1</sup>
- [x] Reviewed build logs for anomalies on GitLab `dev`<sup>1</sup>
- [x] Build passes on GitLab `anvildev`<sup>1</sup>
- [x] Reviewed build logs for anomalies on GitLab `anvildev`<sup>1</sup>
- [x] Deleted PR branch from GitHub
- [x] Deleted PR branch from GitLab `dev`
- [x] Deleted PR branch from GitLab `anvildev`

<sup>1</sup> When pushing the merge commit is skipped due to the PR being
labelled `no sandbox`, the next build triggered by a PR whose merge commit *is*
pushed determines this checklist item.

### Operator (reindex)

- [x] Deleted unreferenced indices in `dev` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Deleted unreferenced indices in `anvildev` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Started reindex in `dev` <sub>or this PR does not require reindexing</sub>
- [x] Started reindex in `anvildev` <sub>or this PR does not require reindexing</sub>
- [x] Checked for and triaged indexing failures in `dev` <sub>or this PR does not require reindexing</sub>
- [x] Checked for and triaged indexing failures in `anvildev` <sub>or this PR does not require reindexing</sub>
- [x] Emptied fail queues in `dev` deployment <sub>or this PR does not require reindexing</sub>
- [x] Emptied fail queues in `anvildev` deployment <sub>or this PR does not require reindexing</sub>

### Operator

- [x] Unassigned PR

## Shorthand for review comments

- `L` line is too long
- `W` line wrapping is wrong
- `Q` bad quotes
- `F` other formatting problem

### CODE REVIEW

**Comment by nadove-ucsc** [comment]
In the partitions indexed in my personal deployment, `disease` and `phenotype` are populated while `phenopacket` is always `null`.

File: src/azul/plugins/metadata/anvil/indexer/transform.py [hunk 1]
```diff
@@ -130,13 +132,15 @@ def field_types(cls) -> FieldTypes:
...
             return BiosampleAggregator()
-        if entity_type == 'datasets':
+        elif entity_type == 'datasets':
             return DatasetAggregator()
-        if entity_type == 'donors':
+        elif entity_type == 'donors':
             return DonorAggregator()
-        if entity_type == 'files':
+        elif entity_type == 'diagnoses':
+            return DiagnosisAggregator()
```
Comment (dsotirho-ucsc) [inline src/azul/plugins/metadata/anvil/indexer/transform.py:140]: Suggestion, move the test of `diagnoses` before `donors` to preserve the alphabetical order of tests

**Review by dsotirho-ucsc — APPROVED** [review]
Approved.

File: src/azul/plugins/metadata/anvil/indexer/transform.py (hunk 1 above)
Comment (hannes-ucsc) [reply src/azul/plugins/metadata/anvil/indexer/transform.py:140]: Hmm, yeah, either append at the end to reflect the fact that it was added later, or insert before the `donors` entry to main the alphabetical ordering.

File: src/azul/plugins/repository/tdr_anvil/__init__.py [hunk 2]
```diff
@@ -255,7 +255,12 @@ def _follow_upstream(self,
         return set.union(
             self._upstream_from_files(source, entities['file']),
             self._upstream_from_biosamples(source, entities['biosample']),
-            # Should we also follow donor.parent_donor?
+            # Diagnoses are technically downstream from donors, but they are
```
Comment (hannes-ucsc) [inline src/azul/plugins/repository/tdr_anvil/__init__.py]: Please request PL slot to discuss this.

File: test/indexer/data/9b78f049-83fc-a99b-83bc-8d40e34265cd.results.json [hunk 3]
```diff
@@ -842,6 +842,7 @@
                         ]
                     }
                 ],
+                "diagnoses": [],
```
Comment (hannes-ucsc) [inline test/indexer/data/9b78f049-83fc-a99b-83bc-8d40e34265cd.results.json]: Looks like we're not getting good test coverage from the existing cans. Can we augment the existing cans or switch to different ones?

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

File: src/azul/plugins/repository/tdr_anvil/__init__.py (hunk 2 above)
Comment (nadove-ucsc) [reply src/azul/plugins/repository/tdr_anvil/__init__.py]: @hannes-ucsc: "Tweak the language of the comment to mention that diagnosis entities would not be discovered at all without a 3rd pass, which would in turn be tantamount to making donors the bundle entity."

File: src/azul/plugins/metadata/anvil/service/response.py [hunk 4]
```diff
@@ -195,6 +198,17 @@ def _non_pivotal_entity(self,
             'dataset_id',
             'title'
         },
+        'diagnoses': {
+            'disease',
+            'phenotype',
+            'phenopacket',
+            'onset_age_unit',
+            'diagnosis_age_unit',
+            # These fields are high cardinality but only appear as inner entities
```
Comment (hannes-ucsc) [inline src/azul/plugins/metadata/anvil/service/response.py]:
```suggestion
            # These fields are of high cardinality but only appear as inner entities
```

Not sure, though.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
I'd prefer to see one commit that replaces the cans (without support for diagnoses) followed by a commit that adds the support for diagnoses so we can see what effect that change has on the canned results.

**Review by hannes-ucsc — APPROVED** [review]
Nice.

**Comment by dsotirho-ucsc** [comment]
Failed the `deploy` job on GitLab `anvil` due to an `IndexExistsAndDiffersException`

https://gitlab.anvil.gi.ucsc.edu/ucsc/azul/-/jobs/9032

```
2023-01-20 21:44:29,596 INFO    MainThread: Creating indices ...
2023-01-20 21:44:29,844 DEBUG   MainThread: Creating ES client [vpc-azul-index-anvilbox-y7lvuxpx7jaryr6rpxubm6ey6u.us-east-1.es.amazonaws.com:443]
2023-01-20 21:44:29,873 INFO    MainThread: Found credentials from IAM Role: azul-gitlab
Traceback (most recent call last):
  File "/builds/ucsc/azul/scripts/reindex.py", line 201, in <module>
    main(sys.argv[1:])
  File "/builds/ucsc/azul/scripts/reindex.py", line 167, in main
    azul.reset_indexer(args.catalogs,
  File "/builds/ucsc/azul/src/azul/azulclient.py", line 468, in reset_indexer
    self.create_all_indices(catalog)
  File "/builds/ucsc/azul/src/azul/azulclient.py", line 378, in create_all_indices
    self.index_service.create_indices(catalog)
  File "/builds/ucsc/azul/src/azul/indexer/index_service.py", line 306, in create_indices
    self._check_index(settings=settings,
  File "/builds/ucsc/azul/src/azul/indexer/index_service.py", line 361, in _check_index
    raise IndexExistsAndDiffersException(key, mappings, index['mappings'])
azul.indexer.index_service.IndexExistsAndDiffersException: ('dynamic_templates', {'numeric_detection': False, 'properties': {'entity_id': {'type': 'text', 'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}}}, 'dynamic_templates': [{'strings_as_text': {'match_mapping_type': 'string', 'mapping': {'type': 'text', 'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}}}}, {'other_types_with_keyword': {'match_mapping_type': '*', 'mapping': {'type': '{dynamic_type}', 'fields': {'keyword': {'type': '{dynamic_type}'}}}}}, {'biosample_age_range': {'path_match': 'contents.biosamples.donor_age_at_collection', 'mapping': {'type': 'double_range'}}}, {'diagnosis_age_range': {'path_match': 'contents.diagnoses.diagnosis_age', 'mapping': {'type': 'double_range'}}}, {'diagnosis_onset_age_range': {'path_match': 'contents.diagnoses.diagnosis_onset_age', 'mapping': {'type': 'double_range'}}}]}, {'dynamic_templates': [{'strings_as_text': {'match_mapping_type': 'string', 'mapping': {'fields': {'keyword': {'ignore_above': 256, 'type': 'keyword'}}, 'type': 'text'}}}, {'other_types_with_keyword': {'match_mapping_type': '*', 'mapping': {'fields': {'keyword': {'type': '{dynamic_type}'}}, 'type': '{dynamic_type}'}}}, {'biosample_age_range': {'path_match': 'contents.biosamples.donor_age_at_collection', 'mapping': {'type': 'double_range'}}}], 'numeric_detection': False, 'properties': {'bundle_deleted': {'type': 'boolean', 'fields': {'keyword': {'type': 'boolean'}}}, 'bundle_uuid': {'type': 'text', 'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}}, 'bundle_version': {'type': 'date', 'fields': {'keyword': {'type': 'date'}}}, 'contents': {'properties': {'activities': {'properties': {'activity_id': {'type': 'text', 'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}}, 'activity_table': {'type': 'text', 'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}}, 'activity_type': {'type': 'text', 'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}}, 'assay_type': {'type': 'text', 'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}}, 'data_modality': {'type': 'text', 'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}}, 'date_created': {'type': 'date', 'fields': {'keyword': {'type': 'date'}}}, 'document_id': {'type': 'text', 'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}}, 'reference_assembly': {'type': 'text', 'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}}, 'source_datarepo_row_ids': {'type': 'text', 'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}}}}, 'biosamples': {'properties': {'anatomical_site': {'type': 'text', 'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}}, 'apriori_cell_type': {'type': 'text', 'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}}, 'biosample_id': {'type': 'text', 'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}}, 'biosample_type': {'type': 'text', 'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}}, 'disease': {'type': 'text', 'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}}, 'document_id': {'type': 'text', 'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}}, 'donor_age_at_collection': {'properties': {'gte': {'type': 'float', 'fields': {'keyword': {'type': 'float'}}}, 'lte': {'type': 'float', 'fields': {'keyword': {'type': 'float'}}}}}, 'donor_age_at_collection_unit': {'type': 'text', 'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}}, 'source_datarepo_row_ids': {'type': 'text', 'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}}}}, 'datasets': {'properties': {'consent_group': {'type': 'text', 'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}}, 'data_modality': {'type': 'text', 'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}}, 'data_use_permission': {'type': 'text', 'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}}, 'dataset_id': {'type': 'text', 'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}}, 'document_id': {'type': 'text', 'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}}, 'owner': {'type': 'text', 'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}}, 'principal_investigator': {'type': 'text', 'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}}, 'registered_identifier': {'type': 'text', 'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}}, 'source_datarepo_row_ids': {'type': 'text', 'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}}, 'title': {'type': 'text', 'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}}}}, 'donors': {'properties': {'document_id': {'type': 'text', 'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}}, 'donor_id': {'type': 'text', 'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}}, 'genetic_ancestry': {'type': 'text', 'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}}, 'organism_type': {'type': 'text', 'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}}, 'phenotypic_sex': {'type': 'text', 'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}}, 'reported_ethnicity': {'type': 'text', 'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}}, 'source_datarepo_row_ids': {'type': 'text', 'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}}}}, 'files': {'properties': {'crc32': {'type': 'text', 'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}}, 'data_modality': {'type': 'text', 'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}}, 'document_id': {'type': 'text', 'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}}, 'drs_path': {'type': 'text', 'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}}, 'file_format': {'type': 'text', 'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}}, 'file_id': {'type': 'text', 'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}}, 'file_md5sum': {'type': 'text', 'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}}, 'file_name': {'type': 'text', 'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}}, 'file_size': {'type': 'long', 'fields': {'keyword': {'type': 'long'}}}, 'file_size_': {'type': 'long', 'fields': {'keyword': {'type': 'long'}}}, 'name': {'type': 'text', 'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}}, 'reference_assembly': {'type': 'text', 'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}}, 'sha256': {'type': 'text', 'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}}, 'size': {'type': 'long', 'fields': {'keyword': {'type': 'long'}}}, 'size_': {'type': 'long', 'fields': {'keyword': {'type': 'long'}}}, 'source_datarepo_row_ids': {'type': 'text', 'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}}, 'uuid': {'type': 'text', 'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}}, 'version': {'type': 'date', 'fields': {'keyword': {'type': 'date'}}}}}}}, 'document_id': {'type': 'text', 'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}}, 'entity_id': {'type': 'text', 'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}}, 'source': {'properties': {'id': {'type': 'text', 'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}}, 'spec': {'type': 'text', 'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}}}}}})
make: *** [Makefile:111: create] Error 1
```

**Comment by hannes-ucsc** [comment]
@danielsotirhos, you checked 

Screen Shot 2023-01-20 at 5 23 43 PM

but didn't actually do that. You HAVE to read each item. There is nothing we can do make this any easier.

The build failure can be fixed by running `make delete` locally with the respective deployment selected (`anvilbox` or `anvildev`). I'm assigning PR back to you. I'll update https://github.com/DataBiosphere/azul/issues/3831 to be more clear about the workaround.

**Pull request merged**